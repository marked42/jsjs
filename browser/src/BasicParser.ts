export class BasicParser {
  private index = 0
  constructor(private input: string) {}

  char() {
    return this.input[this.index]
  }

  next(len?: number) {
    len = len || 1
    this.index += len
  }

  eat(char?: string) {
    if (typeof char === 'string') {
      this.assert(char)
    }

    const text = char ? char : this.char()
    this.next(text ? text.length : 1)
    return text
  }

  assert(char: string) {
    if (!this.input.substring(this.index).startsWith(char)) {
      throw new Error(`unexpected char ${char}`)
    }
  }

  consumeWhitespace() {
    const whitespace = /[\t|\r|\n| ]/
    const startIndex = this.index
    while (whitespace.test(this.char())) {
      this.eat()
    }

    return this.input.substring(startIndex, this.index)
  }

  startsWith(input: string) {
    return this.input.substring(this.index).indexOf(input) === 0
  }

  end() {
    return this.index >= this.input.length
  }

  consumeCharsWhile(predicate: (char: string) => boolean) {
    let text = ''
    while (predicate(this.char()) && !this.end()) {
      text += this.eat()
    }

    return text
  }
}
