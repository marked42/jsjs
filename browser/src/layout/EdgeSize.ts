
export class EdgeSize {
  constructor(
    public left: number,
    public top: number,
    public right: number,
    public bottom: number
  ) { }

  static default() {
    return new EdgeSize(0, 0, 0, 0);
  }
}
