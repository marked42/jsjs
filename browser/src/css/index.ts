import { BasicParser } from '../BasicParser'

class Stylesheet {
  constructor(public rules: Rule[]) {}
}

class Rule {
  constructor(
    public selectors: Selector[],
    public declarations: Declaration[]
  ) {}
}

class Selector {
  constructor(
    public tagName: string,
    public id: string,
    public classes: string[]
  ) {}
}

class Declaration {
  constructor(public name: string, public value: string) {}
}

enum DeclarationValueType {
  Keyword,
  Length,
  Color,
}

class DeclarationValue {
  constructor(public type: DeclarationValueType) {}
}

class DeclarationValueKeyword extends DeclarationValue {
  constructor(public value: string) {
    super(DeclarationValueType.Keyword)
  }
}

class DeclarationValueLength extends DeclarationValue {
  constructor(public value: number, public unit: Unit) {
    super(DeclarationValueType.Length)
  }
}

enum Unit {
  Px,
}

interface Color {
  r: number
  g: number
  b: number
  a: number
}

class DeclarationValueColor extends DeclarationValue {
  constructor(public value: Color) {
    super(DeclarationValueType.Color)
  }
}

export class CSSParser extends BasicParser {
  parseCSS() {}
}

export function parseCSS(input: string) {
  const parser = new CSSParser(input)
  return parser.parseCSS()
}
