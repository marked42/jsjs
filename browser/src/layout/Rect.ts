import { EdgeSize } from './EdgeSize'

export class Rect {
  constructor(
    public x: number,
    public y: number,
    public width: number,
    public height: number
  ) {}

  static default() {
    return new Rect(0, 0, 0, 0)
  }

  expand(edgeSize: EdgeSize) {
    return new Rect(
      this.x - edgeSize.left,
      this.y - edgeSize.top,
      this.width + edgeSize.left + edgeSize.right,
      this.height + edgeSize.top + edgeSize.bottom
    )
  }
}
