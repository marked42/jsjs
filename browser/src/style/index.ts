import { ElementNode, Node, TextNode } from '../html/Node'
import { DeclarationValue, parseCSS, StyleSheet } from '../css'
import { parseHTML } from '../html'
import { MatchedRule } from '../css/MatchedRule'
import { Rule } from '../css/Rule'
import { Selector } from '../css/Selector'
import { Specificity } from '../css/Specificity'

interface PropertyMap {
  [key: string]: DeclarationValue
}

export class StyleNode {
  constructor(
    public node: Node,
    public specifiedValues: PropertyMap,
    public children: StyleNode[]
  ) {}

  lookup(name: string, name2: string, fallback: DeclarationValue) {
    return this.specifiedValues[name] || this.specifiedValues[name2] || fallback
  }

  value(name: string) {
    return this.specifiedValues[name]
  }
}

export function createStyleTree(html: string, css: string) {
  const node = parseHTML(html)
  const stylesheet = parseCSS(css)

  if (node instanceof TextNode) {
    return new StyleNode(node, {}, [])
  }

  return doCreateStyleTree(node, stylesheet)
}

// 类型推导问题 'doCreateStyleTree' implicitly has return type 'any' because it does not have a return type annotation and is referenced directly or indirectly in one of its return expressions.ts(7023)
export function doCreateStyleTree(
  node: Node,
  stylesheet: StyleSheet
): StyleNode {
  if (node instanceof TextNode) {
    return new StyleNode(node, {}, [])
  }

  if (node instanceof ElementNode) {
    return new StyleNode(
      node,
      specifiedValuesForNode(node, stylesheet),
      node.children.map((child) => doCreateStyleTree(child, stylesheet))
    )
  }

  throw new Error(`unexpected node type ${node}`)
}

function specifiedValuesForNode(node: ElementNode, stylesheet: StyleSheet) {
  const matchedRules = matchRulesForNode(node, stylesheet)

  const specifiedValues: PropertyMap = {}
  for (const rule of matchedRules) {
    rule.rule.declarations.forEach((value) => {
      specifiedValues[value.name] = value.value
    })
  }

  return specifiedValues
}

function matchRulesForNode(
  node: ElementNode,
  stylesheet: StyleSheet
): MatchedRule[] {
  const matchedRules: MatchedRule[] = []

  for (const rule of stylesheet.rules) {
    const specificity = nodeMatchedWithRule(node, rule)
    if (specificity) {
      matchedRules.push(new MatchedRule(specificity, rule))
    }
  }

  matchedRules.sort((a, b) => a.specificity.compare(b.specificity))
  return matchedRules
}

function nodeMatchedWithRule(node: ElementNode, rule: Rule) {
  let specificity: Specificity | null = null

  for (const selector of rule.selectors) {
    if (selectorMatchedWithRule(node, selector)) {
      if (!specificity) {
        specificity = Specificity.from(selector)
      } else {
        const newSpecificity = Specificity.from(selector)
        if (specificity.compare(newSpecificity) < 0) {
          specificity = newSpecificity
        }
      }
    }
  }

  return specificity
}

function selectorMatchedWithRule(node: ElementNode, selector: Selector) {
  if (selector.tagName && selector.tagName !== node.name) {
    return false
  }

  if (selector.id && selector.id !== node.attributes['id']) {
    return false
  }

  const classAttributeValue = node.attributes['class']
  const classes: string[] =
    classAttributeValue === undefined ? [] : classAttributeValue.split(' ')
  if (
    selector.classes.length > 0 &&
    !classes.some((klass) => selector.classes.includes(klass))
  ) {
    return false
  }

  return true
}
