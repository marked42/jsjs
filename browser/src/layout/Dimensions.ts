import { EdgeSize } from './EdgeSize'
import { Rect } from './Rect'

export class Dimensions {
  constructor(
    public content: Rect,
    public padding: EdgeSize,
    public border: EdgeSize,
    public margin: EdgeSize
  ) {}

  static default() {
    return new Dimensions(
      Rect.default(),
      EdgeSize.default(),
      EdgeSize.default(),
      EdgeSize.default()
    )
  }

  contentBox() {
    return this.content
  }

  paddingBox() {
    return this.contentBox().expand(this.padding)
  }

  borderBox() {
    return this.paddingBox().expand(this.border)
  }

  marginBox() {
    return this.borderBox().expand(this.margin)
  }
}
