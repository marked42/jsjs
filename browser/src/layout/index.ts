import { TokenEOF } from '../../../c4/Token'
import {
  DeclarationValue,
  DeclarationValueType,
  DeclarationValueKeyword,
} from '../css'
import { StyleNode } from '../style'

export class Rect {
  constructor(
    public x: number,
    public y: number,
    public width: number,
    public height: number
  ) {}

  static default() {
    return new Rect(0, 0, 0, 0)
  }
}

export class EdgeSize {
  constructor(
    public left: number,
    public top: number,
    public right: number,
    public bottom: number
  ) {}

  static default() {
    return new EdgeSize(0, 0, 0, 0)
  }
}

export class Dimensions {
  constructor(
    public content: Rect,
    public padding: EdgeSize,
    public border: EdgeSize,
    public margin: EdgeSize
  ) {}

  static default() {
    return new Dimensions(
      Rect.default(),
      EdgeSize.default(),
      EdgeSize.default(),
      EdgeSize.default()
    )
  }
}

export enum LayoutBoxType {
  Block,
  Inline,
  Anonymous,
}

export class LayoutBox {
  public dimensions = Dimensions.default()
  public children = [] as LayoutBox[]
  constructor(public boxType: LayoutBoxType) {}
}

export function getLayoutBoxType(node: StyleNode) {
  const display = node.specifiedValues['display']

  if (display instanceof DeclarationValueKeyword) {
    if (display.value === 'block') {
      return LayoutBoxType.Block
    } else if (display.value === 'none') {
      return LayoutBoxType.Inline
    } else {
      return LayoutBoxType.Block
    }
  }
  return LayoutBoxType.Block
}

export function getDisplayValue(node: StyleNode) {
  const display = node.specifiedValues['display']

  if (display instanceof DeclarationValueKeyword) {
    if (display.value === 'block') {
      return 'block'
    } else if (display.value === 'none') {
      return 'none'
    } else {
      return 'inline'
    }
  }
  return 'inline'
}

export function buildLayoutTree(node: StyleNode) {
  const display = getDisplayValue(node)
  const root = new LayoutBox(LayoutBoxType.Block)
  switch (display) {
    case 'inline':
      root.boxType = LayoutBoxType.Inline
      break
    case 'block':
      root.boxType = LayoutBoxType.Block
      break
    case 'none':
      throw new Error('display none at root node not allowed')
  }

  node.children.forEach((child) => {
    const childDisplay = getDisplayValue(child)
    switch (childDisplay) {
      case 'block':
        root.children.push(buildLayoutTree(child))
        break
      case 'inline':
        getInlineContainer(root).children.push(buildLayoutTree(child))
      case 'none':
        break
    }
  })

  return root
}

export function getInlineContainer(box: LayoutBox) {
  switch (box.boxType) {
    case LayoutBoxType.Inline:
    case LayoutBoxType.Anonymous:
      return box
    case LayoutBoxType.Block:
      const lastChild = box.children[box.children.length - 1]
      if (lastChild && lastChild.boxType === LayoutBoxType.Anonymous) {
        return lastChild
      } else {
        const anonymousBox = new LayoutBox(LayoutBoxType.Anonymous)

        box.children.push(anonymousBox)
        return anonymousBox
      }
  }
}
