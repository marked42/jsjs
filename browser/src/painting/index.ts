import { LayoutBox, LayoutBoxType } from '../layout'
import { Rect } from '../layout/Rect'
import { Color, DeclarationValueColor } from '../css'

class DisplayCommand {}

class DisplayCommandSolidColor extends DisplayCommand {
  constructor(public color: Color, public rect: Rect) {
    super()
  }
}

export function buildDisplayList(layoutBox: LayoutBox): DisplayCommand[] {
  const displayList = [] as DisplayCommand[]
  renderLayoutBox(displayList, layoutBox)
  return displayList
}

export function renderLayoutBox(list: DisplayCommand[], layoutBox: LayoutBox) {
  renderBackground(list, layoutBox)
  renderBorders(list, layoutBox)

  layoutBox.children.forEach((child) => {
    renderLayoutBox(list, child)
  })
}

export function renderBackground(list: DisplayCommand[], layoutBox: LayoutBox) {
  const backgroundColor = getBoxColor(layoutBox, 'background')
  if (!backgroundColor) {
    return
  }

  list.push(
    new DisplayCommandSolidColor(
      backgroundColor,
      layoutBox.dimensions.borderBox()
    )
  )
}

export function renderBorders(list: DisplayCommand[], layoutBox: LayoutBox) {
  const borderColor = getBoxColor(layoutBox, 'border-color')
  if (!borderColor) {
    return
  }

  const dimensions = layoutBox.dimensions
  const borderBox = layoutBox.dimensions.borderBox()

  // left
  list.push(
    new DisplayCommandSolidColor(
      borderColor,
      new Rect(
        borderBox.x,
        borderBox.y,
        dimensions.border.left,
        borderBox.height
      )
    )
  )

  // right
  list.push(
    new DisplayCommandSolidColor(
      borderColor,
      new Rect(
        borderBox.x +
          dimensions.border.left +
          dimensions.padding.left +
          dimensions.content.width,
        borderBox.y,
        dimensions.border.right,
        borderBox.height
      )
    )
  )

  // top border
  list.push(
    new DisplayCommandSolidColor(
      borderColor,
      new Rect(borderBox.x, borderBox.y, dimensions.border.top, borderBox.width)
    )
  )

  // bottom color
  list.push(
    new DisplayCommandSolidColor(
      borderColor,
      new Rect(
        borderBox.x,
        borderBox.y + borderBox.height - dimensions.border.bottom,
        borderBox.width,
        dimensions.border.bottom
      )
    )
  )
}

export function getBoxColor(layoutBox: LayoutBox, name: string) {
  switch (layoutBox.boxType) {
    case LayoutBoxType.Block:
    case LayoutBoxType.Inline:
      const styleNode = layoutBox.getStyleNode()
      const value = styleNode.value(name)
      if (value) {
        if (value instanceof DeclarationValueColor) {
          return value.value
        } else {
          throw new Error(`style ${name} value is not color, get ${value}|`)
        }
      }
    case LayoutBoxType.Anonymous:
    default:
      break
  }
}

interface CanvasData {
  data: Uint8ClampedArray
  width: number
  height: number
}

export function drawLayoutBoxAsCanvasData(layoutBox: LayoutBox, bounds: Rect) {
  const canvasData = {
    data: new Uint8ClampedArray(4 * bounds.width * bounds.height),
    width: bounds.width,
    height: bounds.height,
  }

  const displayCommandList = buildDisplayList(layoutBox)

  for (const command of displayCommandList) {
    if (command instanceof DisplayCommandSolidColor) {
      applyCommandToCanvasData(command, canvasData)
    }
  }

  return canvasData
}

function applyCommandToCanvasData(
  command: DisplayCommandSolidColor,
  canvasData: CanvasData
) {
  for (let row = 0; row < canvasData.height; row++) {
    for (let col = 0; col < canvasData.width; col++) {
      canvasData.data[row * canvasData.width + col] = command.color.r
      canvasData.data[row * canvasData.width + col + 1] = command.color.g
      canvasData.data[row * canvasData.width + col + 2] = command.color.b
      canvasData.data[row * canvasData.width + col + 3] = command.color.a
    }
  }
}
