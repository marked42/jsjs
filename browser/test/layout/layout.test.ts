import { buildLayoutTree } from '../../src/layout'
import { createStyleTree } from '../../src/style'

describe('box generation', () => {
  it('should generate anonymous boxes', () => {
    const html = `<html><a></a><b></b><c></c><d></d></html>`
    const css = `
		html { display: block; }
		a    { display: block; }
		b, c { display: inline; }
		d    { display: block; }
	`

    expect(buildLayoutTree(createStyleTree(html, css))).toMatchSnapshot()
  })
})
