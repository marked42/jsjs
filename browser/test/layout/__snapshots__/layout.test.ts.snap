// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`box generation should generate anonymous boxes 1`] = `
LayoutBox {
  "boxType": 0,
  "children": Array [
    LayoutBox {
      "boxType": 0,
      "children": Array [],
      "dimensions": Dimensions {
        "border": EdgeSize {
          "bottom": 0,
          "left": 0,
          "right": 0,
          "top": 0,
        },
        "content": Rect {
          "height": 0,
          "width": 0,
          "x": 0,
          "y": 0,
        },
        "margin": EdgeSize {
          "bottom": 0,
          "left": 0,
          "right": 0,
          "top": 0,
        },
        "padding": EdgeSize {
          "bottom": 0,
          "left": 0,
          "right": 0,
          "top": 0,
        },
      },
    },
    LayoutBox {
      "boxType": 2,
      "children": Array [
        LayoutBox {
          "boxType": 1,
          "children": Array [],
          "dimensions": Dimensions {
            "border": EdgeSize {
              "bottom": 0,
              "left": 0,
              "right": 0,
              "top": 0,
            },
            "content": Rect {
              "height": 0,
              "width": 0,
              "x": 0,
              "y": 0,
            },
            "margin": EdgeSize {
              "bottom": 0,
              "left": 0,
              "right": 0,
              "top": 0,
            },
            "padding": EdgeSize {
              "bottom": 0,
              "left": 0,
              "right": 0,
              "top": 0,
            },
          },
        },
        LayoutBox {
          "boxType": 1,
          "children": Array [],
          "dimensions": Dimensions {
            "border": EdgeSize {
              "bottom": 0,
              "left": 0,
              "right": 0,
              "top": 0,
            },
            "content": Rect {
              "height": 0,
              "width": 0,
              "x": 0,
              "y": 0,
            },
            "margin": EdgeSize {
              "bottom": 0,
              "left": 0,
              "right": 0,
              "top": 0,
            },
            "padding": EdgeSize {
              "bottom": 0,
              "left": 0,
              "right": 0,
              "top": 0,
            },
          },
        },
      ],
      "dimensions": Dimensions {
        "border": EdgeSize {
          "bottom": 0,
          "left": 0,
          "right": 0,
          "top": 0,
        },
        "content": Rect {
          "height": 0,
          "width": 0,
          "x": 0,
          "y": 0,
        },
        "margin": EdgeSize {
          "bottom": 0,
          "left": 0,
          "right": 0,
          "top": 0,
        },
        "padding": EdgeSize {
          "bottom": 0,
          "left": 0,
          "right": 0,
          "top": 0,
        },
      },
    },
    LayoutBox {
      "boxType": 0,
      "children": Array [],
      "dimensions": Dimensions {
        "border": EdgeSize {
          "bottom": 0,
          "left": 0,
          "right": 0,
          "top": 0,
        },
        "content": Rect {
          "height": 0,
          "width": 0,
          "x": 0,
          "y": 0,
        },
        "margin": EdgeSize {
          "bottom": 0,
          "left": 0,
          "right": 0,
          "top": 0,
        },
        "padding": EdgeSize {
          "bottom": 0,
          "left": 0,
          "right": 0,
          "top": 0,
        },
      },
    },
  ],
  "dimensions": Dimensions {
    "border": EdgeSize {
      "bottom": 0,
      "left": 0,
      "right": 0,
      "top": 0,
    },
    "content": Rect {
      "height": 0,
      "width": 0,
      "x": 0,
      "y": 0,
    },
    "margin": EdgeSize {
      "bottom": 0,
      "left": 0,
      "right": 0,
      "top": 0,
    },
    "padding": EdgeSize {
      "bottom": 0,
      "left": 0,
      "right": 0,
      "top": 0,
    },
  },
}
`;
